hs_file_types/forms.py

```diff
@@ -0,0 +1,282 @@
+import json
+import jsonschema
+from urllib.request import Request, urlopen
+from urllib.error import URLError, HTTPError
+
+from django import forms
+from .models.model_program import ModelProgramResourceFileType
+from .utils import get_model_program_aggregations
+
+
+class ModelInstanceMetadataValidationForm(forms.Form):
+    def __init__(self, *args, **kwargs):
+        self.user = kwargs.pop('user', None)
+        super(ModelInstanceMetadataValidationForm, self).__init__(*args, **kwargs)
+
+    has_model_output = forms.BooleanField(required=False)
+    executed_by = forms.IntegerField(required=False)
+    executed_by_url = forms.URLField(required=False)
+
+    def clean_executed_by(self):
+        executed_by = self.cleaned_data['executed_by']
+        user_selected_mp_aggr = None
+        # if a model program has been selected then this form would have the id of that aggregation
+        if executed_by is not None and executed_by > 0:
+            user_mp_aggregations = get_model_program_aggregations(self.user)
+            user_selected_mp_aggr = [mp_aggr for mp_aggr in user_mp_aggregations if mp_aggr.id == executed_by]
+            if user_selected_mp_aggr:
+                user_selected_mp_aggr = user_selected_mp_aggr[0]
+            else:
+                self.add_error("executed_by", "You don't have access to the selected model program aggregation")
+        return user_selected_mp_aggr
+
+    def clean_executed_by_url(self):
+        executed_by_url = self.cleaned_data['executed_by_url']
+        if executed_by_url:
+            # check that it's an active url
+            url = Request(executed_by_url)
+            try:
+                urlopen(url)
+            except (URLError, HTTPError):
+                self.add_error("executed_by_url", "Provided URL seems like a broken URL")
+        return executed_by_url
+
+    def update_metadata(self, metadata):
+        executed_by = self.cleaned_data['executed_by']
+        executed_by_url = self.cleaned_data['executed_by_url']
+        if executed_by:
```

Comment:
So you can only have executed_by or executed_by_url.  If both are supplied, then executed_by_url is ignored?  What's driving this?

Reply:
Initially there was a requirement to link to a model program (that leaves outside of hydroshare) by url. That feature was removed later - so no more using `executed_by_url'.